pkg/controller/common.go

```diff
@@ -136,3 +172,349 @@ func cleanUpAzVolumeAttachments(ctx context.Context, client client.Client, azCli
+		return nodeScores[filteredNodes[i]] < nodeScores[filteredNodes[j]]
+	})
+
+	if len(filteredNodes) > numNodes {
+		return filteredNodes[:numNodes], nil
+	}
+	return filteredNodes, nil
+}
+
+func getNodesWithReplica(ctx context.Context, azr azReconciler, volumeName string) ([]string, error) {
+	// get all replica attachments for the given volume
+	volReq, err := createLabelRequirements(azureutils.VolumeNameLabel, volumeName)
+	if err != nil {
+		return nil, err
+	}
+	roleReq, err := createLabelRequirements(azureutils.RoleLabel, string(v1alpha1.ReplicaRole))
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector().Add(*volReq, *roleReq)
+	azVolumeAttachments := &v1alpha1.AzVolumeAttachmentList{}
+	err = azr.getClient().List(ctx, azVolumeAttachments, &client.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := []string{}
+	for _, azVolumeAttachment := range azVolumeAttachments.Items {
+		nodes = append(nodes, azVolumeAttachment.Spec.NodeName)
+	}
+
+	return nodes, nil
+}
+
+func createLabelRequirements(label, value string) (*labels.Requirement, error) {
+	req, err := labels.NewRequirement(label, selection.Equals, []string{value})
+	if err != nil {
+		return nil, err
+	}
+	if req == nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to create Requirement to for label key : (%s) and label value: (%s)", label, value))
+	}
+	return req, nil
+}
+
+func createReplica(ctx context.Context, azr azReconciler, volumeID, node string) error {
+	diskName, err := azureutils.GetDiskNameFromAzureManagedDiskURI(volumeID)
+	if err != nil {
+		return err
+	}
+	volumeName := strings.ToLower(diskName)
+	replicaName := azureutils.GetAzVolumeAttachmentName(volumeName, node)
+
+	_, err = azr.getAzClient().DiskV1alpha1().AzVolumeAttachments(azr.getNamespace()).Create(ctx, &v1alpha1.AzVolumeAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      replicaName,
+			Namespace: azr.getNamespace(),
+			Labels: map[string]string{
+				azureutils.NodeNameLabel:   node,
+				azureutils.VolumeNameLabel: volumeName,
+				azureutils.RoleLabel:       string(v1alpha1.ReplicaRole),
+			},
+		},
+		Spec: v1alpha1.AzVolumeAttachmentSpec{
+			NodeName:         node,
+			VolumeID:         volumeID,
+			UnderlyingVolume: volumeName,
+			RequestedRole:    v1alpha1.ReplicaRole,
+			VolumeContext:    map[string]string{},
+		},
+		Status: v1alpha1.AzVolumeAttachmentStatus{
+			State: v1alpha1.AttachmentPending,
+		},
+	}, metav1.CreateOptions{})
+
+	if err != nil {
+		klog.Infof("Replica AzVolumeAttachment (%s) has been successfully created.", replicaName)
+	}
+	return err
+}
+
+func getNamespacedName(namespace, name string) string {
+	return fmt.Sprintf("%s/%s", namespace, name)
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func NewSharedMap() *SharedMap {
```

Comment:
Did you consider using [`sync.Map`](https://pkg.go.dev/sync#Map) rather than rolling your own?

Reply:
I remember reading somewhere that discouraged uses of `sync.Map`, which was I stayed away from using it in the first place but I don't see any particular reason why `sync.Map` wouldn't work for our controller package.